### PR TITLE
Libc poc

### PR DESCRIPTION
**Pull request by arnetheduck**
This is a proof-of-concept branch of a new way of handling C (and indeed, other system / os / platform / c libraries) - primarily a replacement for posix.nim.

Instead of having a single posix.nim that loosely imports anything posix-related (of an unknown version of posix, sprinkled with some platform-specific extensions here and there), I'd like to suggest a different strategy completely:

* as much of libc (and similar system libraries, msvcrt on windows, etc) as possible is made available in platform specific files, each corresponding to a c header
* platform support is divided into two tiers - abi-compatible and generic
* the objective of the abi-compatible files is to offer up every type, constant and function with their native types, without any conversion or nimification, as is available on that specific platform
* generic types, consts and procs follow the same path as posix.nim does now - best effort, whatever happens to work and be needed, no abi compatbility, fuzzy on what version/standard/etc is actually supported

the branch reuses detect.nim to generate constants, and in a distant future, something like c2nim could maybe be used to do a first cut of types and procs, though c system types tend to be full of unions, raw byte arrays and other nastiness that's difficult to represent accurately in nim, hence doing the conversion by hand might not be a bad option either - once done, it never changes, so there's no need for repeatable automation.

the long term goal of this change is two-fold:
* allow nim code to be compiled standalone, without a c compiler or c header files - this requries an accurate description of the ABI (ie nlvm)
* split up the monolithic posix.nim such that it's easier to include platform-specific features or simply reuse posixy ones - a lower barrier of entry here might encourage reuse rather than slightly incompatible reimport that is very real in the rest of `lib/` today

### CODE REVIEW

**Comment by arnetheduck** [comment]
Look at the commit ed69df1 for a practical view of how this might look - other commits are part of some separate refactoring

File: lib/posix/posix.nim [hunk 1]
```diff
@@ -86,6 +90,10 @@ when defined(linux) and defined(amd64):
 else:
   include posix_other
 
+# Backwards compatibility - would be nice if we could export xxx.*!
```
Comment (zielmicha) [inline lib/posix/posix.nim]: You can export xxx.* by using `export xxx`.

**Comment by zielmicha** [comment]
I like this idea, but this looks like a lot of work. Are you willing to do this to every header for every supported platform?

For example, we need infrastructure to run `detect.nim` on every supported POSIX platform/arch combo.

**Comment by arnetheduck** [comment]
Converting the existing headers is pretty easy, detect.nim does most of the hard work - the two examples I did took no more than few minutes to get the files in the right place and there's not so many of them after all. 

Other maintainers would have to step up to champion their platforms -  the generic, catch-all solution works until they do, no different from the situation today. The point of this exercise is to set an easy and approachable structured and centralized model for adding other platform-specific c / os library stuff such that it can grow over time and piece by piece.

Irrespective of this branch, I'm on my way to submit a patch as well that will run `detect.nim` on the automated builds, and maybe upload the results somewhere - that should take care of the three platforms that receive regular testing, at least

**Comment by krux02** [comment]
Well this PR changes a lot and generally think these changes are in the right direction. But I currently I would prefer if you would release this posix module as a nimble package then we can then reference in the documentation as an alternative to the posix module. As a nimble module you posix module will also be more available to users than as this unmerged PR. When people end up liking it a lot I can imagine to let it replace the posix module from the standard library, but at the moment I don't like to introduce big changes that aren't field tested.

**Comment by arnetheduck** [comment]
Er, it's the opposite of a `posix` module - it's a `c` module. It can be used to implement a `posix` module more cleanly.

I'd consider it a cleanup more than a feature - it's primarily a refactoring that separates the mechanical `importc` parts from any "nimified" API/rebranding. Moving this out would not help the std lib to become any more clean - in fact, the primary user of this module is is the std lib itself - it's not that useful as a nimble package, unless we also move large parts of the std lib out (a thing I'm generally in favor of).

The approach itself is battletested in other languages, it's not a novel idea.

**Pull request closed without merging**